lib/lib.ml

```diff
@@ -2,175 +2,279 @@
  * attachment-converter.ml
  * Copyright (c) 2021 Matt Teichman. All rights reserved.
  * Distributed under the ISC license, see terms at the end of the file.
- *)
+*)
 
 open Prelude
 
 (* library code for attachment converter goes here *)
-
-(* including Owen's demo code from Winter 2021 below *)
-
-open Mrmime
-
-let parse_mail =
-  Angstrom.(parse_string ~consume:All Mail.mail)
-   
-let parse_mail_file fname =
-  readfile fname |> parse_mail
-
-let unpack_root_header = Result.map fst
+module type CONVERT =
+sig
+  type filepath
+  type parsetree
+  type htransform = string -> string
+  type btransform = string -> string
+  val parse : string -> parsetree
+  val amap : htransform -> btransform -> parsetree -> parsetree
+  val acopy : htransform -> btransform -> parsetree -> parsetree
+  val to_string : parsetree -> string
+  val convert : filepath -> (string -> string)
+  val acopy_email : string -> (string -> string) -> string
+end
+
+module Conversion_ocamlnet : CONVERT = struct
+  include Netmime
+  include Netmime_channels
+  include Netchannels
+  include Netstream
+  include Netmime_header
+  include Stdlib.Buffer
+
+
+  type htransform = string -> string
+  type btransform = string -> string
+  type filepath =  String.t
+  type parsetree = Netmime.complex_mime_message
+
+
+  let parse (s : string) =
+    let ch = (new Netstream.input_stream (new Netchannels.input_string s)) in
+    let f = (fun ch -> Netmime_channels.read_mime_message ~multipart_style:`Deep ch) in
+    Netchannels.with_in_obj_channel ch f
+  (* see http://projects.camlcity.org/projects/dl/ocamlnet-4.1.9/doc/html-main/Netmime_tut.html
+     -- I /think/ that with_in_obj_channel should close both the Netchannels and the Netstream input bits,
+     but it's worth keeping an eye on. *)
+
+  let header_to_string (h : mime_header) =
+    let buf = Stdlib.Buffer.create 1024 in
+    Netchannels.with_out_obj_channel
+      (new Netchannels.output_buffer buf)
+      (fun ch -> Netmime_string.write_header ch (h#fields));
+    Stdlib.Buffer.contents buf
+
+  let header_from_string s =
+    new basic_mime_header
+      (Netchannels.with_in_obj_channel
+         (new Netchannels.input_string s)
+         (fun ch -> Netmime_string.read_header (new Netstream.input_stream ch)))
+
+
+  let rec amap f g (tree : parsetree) =
```

Comment:
Just for posterity, the issue is with ocamlnet being picky about line endings in `read_header`. Because those line endings are correct in the actual emails generated by `readpst`, this is really only something we need to be aware of when writing small tests like the one above -- IIRC replacing the `(const "application/json")` with `(const "Content-Type: application/json\r\n\r\n")` was enough to fix this behavior without changing `amap`